File: movemork/Data/Repositories/PropertyRepository.swift
//
//  PropertyRepository.swift
//  movemork
//
//  MoveMark — Property and room data access.
//

import Foundation
import Supabase

struct PropertyRow: Codable, Identifiable {
    let id: UUID
    let userId: UUID
    var title: String
    var addressLine1: String
    var addressLine2: String?
    var city: String
    var provinceState: String
    var postalCode: String
    var country: String
    var landlordName: String?
    var landlordEmail: String?
    var landlordPhone: String?
    var depositAmount: Double?
    var rentAmount: Double?
    var moveInDate: String?
    var leaseStart: String?
    var leaseEnd: String?
    var createdAt: String?

    enum CodingKeys: String, CodingKey {
        case id
        case userId = "user_id"
        case title
        case addressLine1 = "address_line_1"
        case addressLine2 = "address_line_2"
        case city
        case provinceState = "province_state"
        case postalCode = "postal_code"
        case country
        case landlordName = "landlord_name"
        case landlordEmail = "landlord_email"
        case landlordPhone = "landlord_phone"
        case depositAmount = "deposit_amount"
        case rentAmount = "rent_amount"
        case moveInDate = "move_in_date"
        case leaseStart = "lease_start"
        case leaseEnd = "lease_end"
        case createdAt = "created_at"
    }
}

struct RoomRow: Codable, Identifiable {
    let id: UUID
    let propertyId: UUID
    var name: String
    var sortOrder: Int
    var createdAt: String?

    enum CodingKeys: String, CodingKey {
        case id
        case propertyId = "property_id"
        case name
        case sortOrder = "sort_order"
        case createdAt = "created_at"
    }
}

struct PropertyRepository {

    func fetchProperties(userId: UUID) async throws -> [PropertyRow] {
        try await supabase
            .from("properties")
            .select()
            .eq("user_id", value: userId)
            .execute()
            .value
    }

    func createProperty(_ row: PropertyRow) async throws -> PropertyRow {
        #if DEBUG
        if let encoded = try? JSONEncoder().encode(row),
           let json = String(data: encoded, encoding: .utf8) {
            print("📦 Property insert payload (JSON): \(json)")
        } else {
            print("📦 Property insert payload: \(row)")
        }
        #endif
        do {
            let created: PropertyRow = try await supabase
                .from("properties")
                .insert(row)
                .select()
                .single()
                .execute()
                .value
            #if DEBUG
            print("✅ Property created: \(created.id)")
            #endif
            return created
        } catch {
            #if DEBUG
            print("❌ createProperty failed: \(error)")
            #endif
            throw error
        }
    }

    func updateProperty(_ row: PropertyRow) async throws {
        struct PropertyPatch: Encodable {
            let title: String
            let addressLine1: String
            let addressLine2: String?
            let city: String
            let provinceState: String
            let postalCode: String
            let country: String
            let landlordName: String?
            let landlordEmail: String?
            let landlordPhone: String?
            let depositAmount: Double?
            let rentAmount: Double?
            let moveInDate: String?
            let leaseStart: String?
            let leaseEnd: String?

            enum CodingKeys: String, CodingKey {
                case title
                case addressLine1 = "address_line_1"
                case addressLine2 = "address_line_2"
                case city
                case provinceState = "province_state"
                case postalCode = "postal_code"
                case country
                case landlordName = "landlord_name"
                case landlordEmail = "landlord_email"
                case landlordPhone = "landlord_phone"
                case depositAmount = "deposit_amount"
                case rentAmount = "rent_amount"
                case moveInDate = "move_in_date"
                case leaseStart = "lease_start"
                case leaseEnd = "lease_end"
            }
        }

        let patch = PropertyPatch(
            title: row.title,
            addressLine1: row.addressLine1,
            addressLine2: row.addressLine2,
            city: row.city,
            provinceState: row.provinceState,
            postalCode: row.postalCode,
            country: row.country,
            landlordName: row.landlordName,
            landlordEmail: row.landlordEmail,
            landlordPhone: row.landlordPhone,
            depositAmount: row.depositAmount,
            rentAmount: row.rentAmount,
            moveInDate: row.moveInDate,
            leaseStart: row.leaseStart,
            leaseEnd: row.leaseEnd
        )

        try await supabase
            .from("properties")
            .update(patch)
            .eq("id", value: row.id)
            .execute()
    }

    func fetchRooms(propertyId: UUID) async throws -> [RoomRow] {
        try await supabase
            .from("rooms")
            .select()
            .eq("property_id", value: propertyId)
            .order("sort_order")
            .execute()
            .value
    }

    func insertRoom(_ row: RoomRow) async throws -> RoomRow {
        try await supabase
            .from("rooms")
            .insert(row)
            .select()
            .single()
            .execute()
            .value
    }

    func insertDefaultRooms(propertyId: UUID, userId: UUID) async throws {
        let names = [
            "Kitchen", "Living Room", "Primary Bedroom", "Bedroom 2",
            "Bathroom", "Bathroom 2", "Hallway", "Laundry",
            "Storage", "Balcony/Patio"
        ]

        let rooms = names.enumerated().map { index, name in
            RoomRow(
                id: UUID(),
                propertyId: propertyId,
                name: name,
                sortOrder: index
            )
        }

        try await supabase
            .from("rooms")
            .insert(rooms)
            .execute()
    }
}

File: movemark-android/app/src/main/java/com/surensureshkumar/movemark/data/property/PropertyRepository.kt
package com.surensureshkumar.movemark.data.property

import com.surensureshkumar.movemark.data.models.CreatePropertyInput
import com.surensureshkumar.movemark.data.models.PropertyRow
import com.surensureshkumar.movemark.data.models.PropertySnapshot
import com.surensureshkumar.movemark.data.models.RoomRow
import io.github.jan.supabase.SupabaseClient
import io.github.jan.supabase.postgrest.from
import io.github.jan.supabase.postgrest.postgrest
import io.github.jan.supabase.postgrest.query.Order
import io.github.jan.supabase.postgrest.rpc
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import java.text.SimpleDateFormat
import java.util.Locale
import java.util.UUID
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class PropertyRepository @Inject constructor(
    private val client: SupabaseClient,
) {
    private val dbDate = SimpleDateFormat("yyyy-MM-dd", Locale.US)
    private companion object {
        const val MAX_PROPERTIES = 300L
    }

    // Decoded independently of the client's default (strict) serializer: the RPC's jsonb payload
    // mirrors whole table rows, and `ignoreUnknownKeys` keeps this resilient to future columns
    // the migration/schema adds that these Kotlin row models haven't been updated to include yet
    // (a decode failure on one unrelated new column would otherwise break the entire snapshot).
    private val snapshotJson = Json { ignoreUnknownKeys = true }

    /**
     * Single-round-trip replacement for the old rooms/inspections/inspection_items/evidence_files/
     * tags/documents/maintenance_issues fan-out. See `get_property_snapshot` in
     * supabase/migrations/20260719000001_property_snapshot_thumbnails_realtime.sql — SECURITY
     * INVOKER, so this still runs under the same per-table RLS as the direct `.select()` calls.
     */
    suspend fun fetchPropertySnapshot(propertyId: UUID): PropertySnapshot =
        decodeSnapshot(fetchPropertySnapshotJson(propertyId))

    /**
     * Same RPC call as [fetchPropertySnapshot], but returns the raw jsonb response string so
     * callers (see [PropertyHydrator]) can persist it verbatim to the offline cache and decode it
     * back with [decodeSnapshot] later, without a lossy re-serialize round trip.
     */
    suspend fun fetchPropertySnapshotJson(propertyId: UUID): String {
        val result = client.postgrest.rpc(
            "get_property_snapshot",
            buildJsonObject { put("p_property_id", propertyId.toString()) },
        )
        return result.data
    }

    fun decodeSnapshot(json: String): PropertySnapshot =
        snapshotJson.decodeFromString(PropertySnapshot.serializer(), json)

    suspend fun fetchProperties(userId: UUID): List<PropertyRow> =
        client.from("properties")
            .select {
                filter { eq("user_id", userId.toString()) }
                // Explicit order (previously implicit/unspecified) plus a cap so a long-time user's
                // property history can't grow this into an unbounded list; oldest-first preserves
                // `fetched.first()`'s existing "earliest property" default-selection behavior.
                order("created_at", Order.ASCENDING)
                limit(MAX_PROPERTIES)
            }
            .decodeList()

    suspend fun fetchRooms(propertyId: UUID): List<RoomRow> =
        client.from("rooms")
            .select {
                filter { eq("property_id", propertyId.toString()) }
                order("sort_order", Order.ASCENDING)
            }
            .decodeList()

    suspend fun createProperty(input: CreatePropertyInput, userId: UUID): PropertyRow {
        val id = UUID.randomUUID()
        val row = PropertyRow(
            id = id.toString(),
            userId = userId.toString(),
            title = input.titleValue,
            addressLine1 = input.addressLine1.trim().ifEmpty { "Address" },
            addressLine2 = input.unit.trim().takeIf { it.isNotEmpty() },
            city = input.city.trim().ifEmpty { "Unknown" },
            provinceState = input.region.trim().ifEmpty { "Unknown" },
            postalCode = input.postalCode.trim().ifEmpty { "00000" },
            country = input.country.trim().ifEmpty { "CA" },
            landlordName = input.landlordName.trim().takeIf { it.isNotEmpty() },
            landlordEmail = input.landlordEmail.trim().takeIf { it.isNotEmpty() },
            landlordPhone = input.landlordPhone.trim().takeIf { it.isNotEmpty() },
            moveInDate = dbDate.format(java.util.Date()),
        )
        return client.from("properties")
            .insert(row) { select() }
            .decodeSingle()
    }

    suspend fun insertDefaultRooms(propertyId: UUID) {
        val names = listOf(
            "Kitchen", "Living Room", "Primary Bedroom", "Bedroom 2",
            "Bathroom", "Bathroom 2", "Hallway", "Laundry",
            "Storage", "Balcony/Patio",
        )
        val rooms = names.mapIndexed { index, name ->
            RoomRow(
                id = UUID.randomUUID().toString(),
                propertyId = propertyId.toString(),
                name = name,
                sortOrder = index,
            )
        }
        client.from("rooms").insert(rooms) {
            select()
        }
    }

    suspend fun insertRoom(propertyId: UUID, name: String, sortOrder: Int): RoomRow {
        val row = RoomRow(
            id = UUID.randomUUID().toString(),
            propertyId = propertyId.toString(),
            name = name,
            sortOrder = sortOrder,
        )
        return client.from("rooms").insert(row) { select() }.decodeSingle()
    }
}
